src/migration.rs

```diff
@@ -0,0 +1,111 @@
+//! TODO
+
+use bevy::{
+    ecs::{
+        entity::Entity,
+        reflect::{AppTypeRegistry, ReflectComponent},
+        resource::Resource,
+        system::{Commands, Query, Res, ResMut, RunSystemOnce},
+        world::World,
+    },
+    platform::collections::HashMap,
+    prelude::{Deref, DerefMut},
+    reflect::{FromType, Reflect, prelude::ReflectDefault},
+};
+use std::{
+    any::{Any, TypeId},
+    sync::Arc,
+};
+
+/// TODO
+pub trait HotPatchMigrate: Any + Reflect + Default {
+    /// TODO
+    fn current_type_id() -> TypeId;
+}
+
+/// TODO
+#[derive(Clone)]
+pub struct ReflectHotPatchMigrate(pub Arc<dyn Fn() -> TypeId + Sync + Send + 'static>);
+
+impl<T: HotPatchMigrate> FromType<T> for ReflectHotPatchMigrate {
+    fn from_type() -> Self {
+        Self(Arc::new(T::current_type_id))
+    }
+}
+
+#[derive(Resource, Default, Deref, DerefMut)]
+pub(crate) struct ComponentMigrations(
+    HashMap<TypeId, Arc<dyn Fn() -> TypeId + Sync + Send + 'static>>,
```

Comment:
I believe Bevy exports a `TypeIdMap` which is supposed to be a bit more optimized.